dev/tools/merge-pr.sh

```diff
@@ -1,50 +1,126 @@
+  exit 1
+fi
+
+if [[ "$1" =~ ^[0-9][0-9]*$ ]]; then
+  PR=$1
+else
+  error $1 is not a number
+  exit 1
+fi
+
+# Fetching PR metadata
+
+TITLE=$(curl -s "$API/pulls/$PR" | jq -r '.title')
+info title for PR $PR is ${BLUE}$TITLE
 
 BASE_BRANCH=$(curl -s "$API/pulls/$PR" | jq -r '.base.label')
+info PR $PR targets branch ${BLUE}$BASE_BRANCH
+
+CURRENT_LOCAL_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+info you are merging in ${BLUE}$CURRENT_LOCAL_BRANCH
+
+REMOTE=$(git config --get "branch.$CURRENT_LOCAL_BRANCH.remote")
+info remote for $CURRENT_LOCAL_BRANCH is ${BLUE}$REMOTE
 
+info fetching from $REMOTE the PR
+git remote update "$REMOTE"
+if ! git ls-remote "$REMOTE" | grep pull >/dev/null; then
+  error remote $REMOTE is not configured to fetch pull requests
+  error run: git config remote.$REMOTE.fetch +refs/pull/*/head:refs/remotes/$REMOTE/pr/*
+  exit 1
+fi
+git fetch "$REMOTE" "refs/pull/$PR/head"
 COMMIT=$(git rev-parse FETCH_HEAD)
-STATUS=$(curl -s "$API/commits/$COMMIT/status" | jq -r '.state')
+info commit for for PR $PR is ${BLUE}$COMMIT
+
+# Sanity check: merge to a different branch
 
 if [ "$BASE_BRANCH" != "coq:$CURRENT_LOCAL_BRANCH" ]; then
-  echo "Wrong base branch"
-  read -p "Bypass? [y/N] " -n 1 -r
-  echo
-  if [[ ! $REPLY =~ ^[Yy]$ ]]
-  then
-    exit 1
-  fi
+  echo "error: PR requests merge in ${BLUE}$BASE_BRANCH${BLACK} but you are merging in ${BLUE}$CURRENT_LOCAL_BRANCH"
+  ask_confirmation
 fi;
 
+# Sanity check: CI failed
+
+STATUS=$(curl -s "$API/commits/$COMMIT/status" | jq -r '.state')
+
 if [ "$STATUS" != "success" ]; then
-  echo "CI status is \"$STATUS\""
-  read -p "Bypass? [y/N] " -n 1 -r
-  echo
-  if [[ ! $REPLY =~ ^[Yy]$ ]]
-  then
-    exit 1
-  fi
+  error CI failure on ${BLUE}$(curl -s "$API/commits/$COMMIT/status" |
+    jq -r -c '.statuses|map(select(.state == "failure"))|map(.context)')
+  ask_confirmation
 fi;
 
-git merge -S --no-ff FETCH_HEAD -m "Merge PR #$PR: $(curl -s "$API/pulls/$PR" | jq -r '.title')" -e
+# Sanity check: has labels named "needs:"
+
+NEEDS_LABELS=$(curl -s "$API/pulls/$PR" | jq -rc '.labels | map(select(.name | match("needs:"))) | map(.name)')
+if [ "$NEEDS_LABELS" != "[]" ]; then
+  error needs:something labels still present: ${BLUE}$NEEDS_LABELS
+  ask_confirmation
+fi
+
+# Sanity check: user.signingkey
+if [ -z "$(git config user.signingkey)" ]; then
+  warning: git config user.signingkey is empty
+  warning: gpg will guess a key out of your git config user.* data
+fi
+
+info merging
+git merge -v -S --no-ff FETCH_HEAD -m "Merge PR #$PR: $TITLE" -e
 
 # TODO: improve this check
-if ! git diff --quiet "$REMOTE/$CURRENT_LOCAL_BRANCH" -- dev/ci; then
-  echo "******************************************"
-  echo "** WARNING: does this PR have overlays? **"
-  echo "******************************************"
+if ! git diff --quiet "$REMOTE/$CURRENT_LOCAL_BRANCH" -- dev/ci/user-overlays; then
+  warning "this PR may have overlays (sorry the check is not perfect)"
```

Comment:
What is the imperfection exactly?

Reply:
We check that something changes in that directory, not that a new overlay is present.

I think the check should be refined to also print the overlays, so that one has a list of links to click for the tasks that are recommended (two lines below)

Reply:
It could be cleaning overlays or something like this. Not worth to mention the imperfection IMHO.

Reply:
I think we can leave with it being mentioned.

Reply:
You meant "live" I guess.

Reply:
Yes, I'm thinking a bit too much about leaving for the week-end probably :)